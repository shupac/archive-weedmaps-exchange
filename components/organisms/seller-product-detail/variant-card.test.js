import { shallow } from 'enzyme';
import mockProductDetails from 'mocks/seller-product-details';
import mockZones from 'mocks/zones';

import VariantCard from './variant-card';

function setup(newProps) {
  const variant = mockProductDetails.product.variants[0];

  const props = {
    variant,
    zones: mockZones,
    onUpdate: jest.fn(),
    onDelete: jest.fn(),
    errors: {},
    touched: {},
    handleBlur: jest.fn(),
    namePath: 'product.variants',
    index: 0,
  };

  const component = <VariantCard {...props} {...newProps} />;
  const wrapper = shallow(component);
  const instance = wrapper.instance();

  const formData = {
    allocations: variant.allocations,
    errors: {},
    touched: {},
    handleBlur: jest.fn(),
  };
  const Allocations = instance.renderAllocations(formData);
  const arrayHelpers = {
    push: jest.fn(),
    replace: jest.fn(),
    remove: jest.fn(),
  };
  const allocationsWrapper = shallow(<Allocations {...arrayHelpers} />);
  return { wrapper, props, instance, allocationsWrapper, arrayHelpers };
}

describe('Seller Product Details Variant Card', () => {
  it('should render the component', () => {
    const { wrapper, allocationsWrapper } = setup();
    expect(wrapper.exists()).toEqual(true);
    expect(allocationsWrapper.exists()).toEqual(true);
  });

  it('should update the name', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').first();
    const event = { target: { value: 'foo' } };
    select.props().onChange(event);
    expect(props.onUpdate).toHaveBeenCalledWith({ name: 'foo' });
  });

  it('should update the sku', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').at(1);
    const event = { target: { value: 'foo' } };
    select.props().onChange(event);
    expect(props.onUpdate).toHaveBeenCalledWith({ sku: 'foo' });
  });

  it('should update the size', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').at(2);
    const event = { target: { value: '1' } };
    select.props().onChange(event);
    expect(props.onUpdate).toHaveBeenCalledWith({ size: '1' });
  });

  it('should render unit option name', () => {
    const { wrapper } = setup();
    const select = wrapper.find('StyledSelect');
    const option = { value: 'foo', text: 'bar' };
    expect(select.props().itemToString(option)).toEqual(option.text);
  });

  it('should update the zone', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('StyledSelect');
    const option = { value: 'foo', text: 'bar' };
    select.props().onChange(option);
    expect(props.onUpdate).toHaveBeenCalledWith({ unit: 'foo' });
  });

  it('should handle add allocation', () => {
    const { allocationsWrapper, arrayHelpers } = setup();
    allocationsWrapper
      .find('AddAllocation')
      .first()
      .simulate('click');
    expect(arrayHelpers.push).toHaveBeenCalled();
  });

  it('should handle update allocation', () => {
    const { allocationsWrapper, arrayHelpers, props } = setup();
    allocationsWrapper
      .find('AllocationRow')
      .first()
      .props()
      .onUpdate({ foo: 'bar' });
    expect(arrayHelpers.replace).toHaveBeenCalledWith(0, {
      ...props.variant.allocations[0],
      foo: 'bar',
    });
  });

  it('should handle delete allocation', () => {
    const { allocationsWrapper, arrayHelpers } = setup();
    allocationsWrapper
      .find('AllocationRow')
      .first()
      .props()
      .onDelete();
    expect(arrayHelpers.remove).toHaveBeenCalledWith(0);
  });

  it('should calculate available zones', () => {
    const { wrapper, instance, props } = setup();
    props.variant.allocations.push({
      id: 'abc',
      active: false,
      amount: '',
      currency: 'usd',
      price: '',
      isNew: true,
    });
    wrapper.setProps(props);
    expect(instance.getAvailableZones()).toMatchSnapshot();
  });

  it('should handle onBlur for the name', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').first();
    const event = { target: { foo: 'bar' } };
    select.props().onBlur(event);
    expect(props.handleBlur).toHaveBeenCalledWith(event);
  });

  it('should handle onBlur for the sku', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').at(1);
    const event = { target: { foo: 'bar' } };
    select.props().onBlur(event);
    expect(props.handleBlur).toHaveBeenCalledWith(event);
  });

  it('should handle onBlur for the size', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('FormInput').at(2);
    const event = { target: { foo: 'bar' } };
    select.props().onBlur(event);
    expect(props.handleBlur).toHaveBeenCalledWith(event);
  });

  it('should handle onBlur for the unit', () => {
    const { wrapper, props } = setup();
    const select = wrapper.find('StyledSelect');
    select.props().onBlur();
    expect(props.handleBlur).toHaveBeenCalledWith({
      target: {
        name: 'product.variants.0.unit',
      },
    });
  });

  it('should show errors', () => {
    const { wrapper } = setup({
      errors: {
        name: 'Test error',
        sku: 'Test error',
        size: 'Test error',
        unit: 'Test error',
      },
      touched: {
        name: true,
        sku: true,
        size: true,
        unit: true,
      },
    });
    expect(wrapper.find('FormInputError').length).toEqual(4);
  });
});
