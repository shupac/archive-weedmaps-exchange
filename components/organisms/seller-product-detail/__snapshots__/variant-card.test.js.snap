// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seller Product Details Variant Card should calculate available zones 1`] = `
Array [
  Object {
    "cId": "0ccf7a68-8f23-4dad-91bb-8d93f9d7cf29",
    "color": "#4A90E2",
    "id": "0ccf7a68-8f23-4dad-91bb-8d93f9d7cf29",
    "name": "NorCal",
    "regions": Array [],
  },
  Object {
    "cId": "89cfc0ca-4410-49e2-af3a-f74808926bbd",
    "color": "#D0021B",
    "id": "89cfc0ca-4410-49e2-af3a-f74808926bbd",
    "name": "Andrew's Zone #3 3449 Updated 5",
    "regions": Array [
      Object {
        "id": "2411d92b-d816-4d63-be75-d1a526cb84b3",
        "name": "Central California",
        "wmRegionId": 4,
      },
      Object {
        "id": "f72ef8f0-3ba0-40d7-b328-613ab851bc26",
        "name": "Oregon",
        "wmRegionId": 5,
      },
      Object {
        "id": "fd167cb8-5be3-4c5e-9207-1e78b771c11d",
        "name": "Arkansas",
        "wmRegionId": 6,
      },
      Object {
        "id": "edd2ef7e-00b4-444d-b3e3-996ccbaa7b95",
        "name": "Texas",
        "wmRegionId": 7,
      },
    ],
  },
]
`;
